x/virtualgroup/keeper/keeper.go

```diff
@@ -0,0 +1,684 @@
+		for i, spID := range gvg.SecondarySpIds {
+			if spID == secondarySPID {
+				gvg.SecondarySpIds[i] = successorSPID
+				origin := k.MustGetGVGStatisticsWithinSP(ctx, secondarySPID)
+				successor := k.MustGetGVGStatisticsWithinSP(ctx, successorSPID)
+				origin.SecondaryCount--
+				successor.SecondaryCount++
+				k.SetGVGStatisticsWithSP(ctx, origin)
+				k.SetGVGStatisticsWithSP(ctx, successor)
+				k.SetGVG(ctx, gvg)
+				break
+			}
+		}
+	}
+	return nil
+}
+
+func (k Keeper) GetOrCreateGVGStatisticsWithinSP(ctx sdk.Context, spID uint32) *types.GVGStatisticsWithinSP {
+	store := ctx.KVStore(k.storeKey)
+
+	ret := &types.GVGStatisticsWithinSP{
+		StorageProviderId: spID,
+		SecondaryCount:    0,
+	}
+	bz := store.Get(types.GetGVGStatisticsWithinSPKey(spID))
+	if bz == nil {
+		return ret
+	}
+
+	k.cdc.MustUnmarshal(bz, ret)
+	return ret
+}
+
+func (k Keeper) GetGVGStatisticsWithinSP(ctx sdk.Context, spID uint32) (*types.GVGStatisticsWithinSP, bool) {
+	store := ctx.KVStore(k.storeKey)
+
+	bz := store.Get(types.GetGVGStatisticsWithinSPKey(spID))
+	if bz == nil {
+		return nil, false
+	}
+
+	var ret types.GVGStatisticsWithinSP
+	k.cdc.MustUnmarshal(bz, &ret)
+	return &ret, true
+}
+
+func (k Keeper) MustGetGVGStatisticsWithinSP(ctx sdk.Context, spID uint32) *types.GVGStatisticsWithinSP {
+	store := ctx.KVStore(k.storeKey)
+
+	bz := store.Get(types.GetGVGStatisticsWithinSPKey(spID))
+	if bz == nil {
+		panic("must get gvg statistics within sp")
+	}
+
+	var ret types.GVGStatisticsWithinSP
+	k.cdc.MustUnmarshal(bz, &ret)
+	return &ret
+}
+
+func (k Keeper) SetGVGStatisticsWithSP(ctx sdk.Context, gvgsStatisticsWithinSP *types.GVGStatisticsWithinSP) {
+	store := ctx.KVStore(k.storeKey)
+
+	bz := k.cdc.MustMarshal(gvgsStatisticsWithinSP)
+
+	store.Set(types.GetGVGStatisticsWithinSPKey(gvgsStatisticsWithinSP.StorageProviderId), bz)
+}
+
+func (k Keeper) BatchSetGVGStatisticsWithinSP(ctx sdk.Context, gvgsStatisticsWithinSP []*types.GVGStatisticsWithinSP) {
+	for _, g := range gvgsStatisticsWithinSP {
+		k.SetGVGStatisticsWithSP(ctx, g)
+	}
+}
+
+func (k Keeper) IsStorageProviderCanExit(ctx sdk.Context, spID uint32) error {
+	store := ctx.KVStore(k.storeKey)
+
+	prefixStore := prefix.NewStore(store, types.GetGVGFamilyPrefixKey(spID))
+	iter := prefixStore.Iterator(nil, nil)
+	if iter.Valid() {
+		var family types.GlobalVirtualGroupFamily
+		k.cdc.MustUnmarshal(iter.Value(), &family)
+		return types.ErrSPCanNotExit.Wrapf("not swap out from all the family, key: %s", family.String())
+	}
+
+	gvgStat, found := k.GetGVGStatisticsWithinSP(ctx, spID)
+	if found && gvgStat.SecondaryCount != 0 {
+		return types.ErrSPCanNotExit.Wrapf("not swap out from all the gvgs, remain: %d", gvgStat.SecondaryCount)
+	}
+	return nil
+}
+
+func (k Keeper) RebindingGVGsToBucket(ctx sdk.Context, bucketID math.Uint, dstSP *sptypes.StorageProvider, newLVGToGVGMappings []*storagetypes.LVGToGVGMapping) error {
+	gvgsBindingOnBucket, found := k.GetGVGsBindingOnBucket(ctx, bucketID)
```

Comment:
How can  `LVGToGVGMapping`  contain `LocalVirtualGroupId`,  I think  `LocalVirtualGroupId` is generated on chain.